onnx/defs/nn/defs.cc

```diff
@@ -222,11 +222,16 @@ std::function<void(OpSchema&)> PoolOpSchemaGenerator(
  ```
  if ceil_mode is enabled `pad_shape[i]` is the sum of pads along axis `i`.
 
- `auto_pad` is a DEPRECATED attribute. If you are using them currently, the output spatial shape will be following:
+ `auto_pad` is a DEPRECATED attribute. If you are using them currently, the output spatial shape will be following when ceil_mode is enabled:
```

Comment:
I found ceil mode is specific to Pytorch where SAME is to TF. Maybe we shall not link equivalence them?

Reply:
> ceil mode is specific to Pytorch where SAME is to TF.

Yes, initially there is only auto_pad in ONNX pooling family operators and later ONNX brought ceil_mode from PyTorch. When introducing ceil_mode, the combination behavior of these two attributes (auto_pad+ceil_mode) are not specified and caused confusion: https://github.com/onnx/onnx/issues/3314. Since it's an undefined behavior in ONNX-level, I am slightly inclined to make ONNX definition clear as current ONNX Runtime's behavior: honor ceil_mode when using auto_pad, to prevent runtime behavior changes, although in practice there shouldn't any real-world models using them (auto_pad+ceil_mode) simultaneously since one is from PyTorch and another is from TensorFlow. It seems fine to have this combination, as long as the document is clear.

Another option is making ONNX forbid using auto_pad and ceil_mode at the same time, but I am not 100% sure whether it would break backward compatibility in runtime.

Reply:
Good point! Thank you for this clarification. It is definitely better to explicitly state the behavior of combination of auto_pad with ceil_mode. I am in the mean time preparing a PR to add to this mix another extra dilation input. In my case I was planning to avoid the interplay between dilation and auto_pad.

Reply:
According to TF documentation: https://www.tensorflow.org/api_docs/python/tf/keras/layers/MaxPool1D#args
"same" means padding to make output size the same as input. "valid" means no padding. Sounds like "same" is close to ceil_mode which is opposite to the changed doc here.
Here is my experiment:
>>> x = tf.constant([1., 2., 3., 4., 0.])
>>> x = tf.reshape(x, [1, 5, 1])
>>> max_pool_1d = tf.keras.layers.MaxPooling1D(pool_size=2, strides=4, padding='same')
>>> max_pool_1d(x)
<tf.Tensor: shape=(1, 2, 1), dtype=float32, numpy=
array([[[2.],
        [0.]]], dtype=float32)>
which matches:
>>> import torch
>>> from torch import nn
>>> m = nn.MaxPool1d(2, stride=4, padding=0, dilation=1, ceil_mode=True)
>>> input = torch.from_numpy(np.array([[[1, 2, 3, 4, 0]]])).float()
>>> output = m(input)
>>> output
tensor([[[2., 0.]]])


on the other hand:
>>> x = tf.constant([1., 2., 3., 4., 0.])
>>> x = tf.reshape(x, [1, 5, 1])
>>> max_pool_1d = tf.keras.layers.MaxPooling1D(pool_size=2, strides=4, padding='same')
>>> max_pool_1d(x)
<tf.Tensor: shape=(1, 2, 1), dtype=float32, numpy=
array([[[2.],
        [0.]]], dtype=float32)>

which matches:
>>> m = nn.MaxPool1d(2, stride=4, padding=0, dilation=1, ceil_mode=True)
>>> input = torch.from_numpy(np.array([[[1, 2, 3, 4, 0]]])).float()
>>> output = m(input)
>>> output
tensor([[[2., 0.]]])